### PR TITLE
[stdlib] String : RangeReplaceableCollection & BidirectionalCollection

### PR DESCRIPTION
**Pull request by moiseev**
rdar://problem/17888632

### CODE REVIEW

**Comment by moiseev** [comment]
@swift-ci Please test and merge

**Review by airspeedswift — APPROVED** [review]

**Comment by airspeedswift** [comment]
@swift-ci Please test

**Comment by airspeedswift** [comment]
@swift-ci Please benchmark

**Comment by airspeedswift** [comment]
Looks like we need the horrible hack to avoid shadowing by `self.min` and `self.max`...

File: stdlib/public/SDK/Foundation/NSStringAPI.swift [hunk 1]
```diff
@@ -636,7 +636,8 @@ extension String {
   public func getCString(
     _ buffer: inout [CChar], maxLength: Int, encoding: Encoding
   ) -> Bool {
-    return _ns.getCString(&buffer, maxLength: min(buffer.count, maxLength),
+    return _ns.getCString(&buffer,
```
Comment (dabrahams) [inline stdlib/public/SDK/Foundation/NSStringAPI.swift]: Please, no endline layout (all opening delimiters whose matching close does not fit on the same line should be followed  by newline)

**Comment by airspeedswift** [comment]
@swift-ci Please test

**Comment by airspeedswift** [comment]
@swift-ci Please benchmark

**Comment by airspeedswift** [comment]
Need a long-term strategy for how to handle the `min` and `max` awfulness... either documentation explaining, or put deprecation warnings on their use nudging towards `Swift.max`, or fix the compiler (which probably won't happen for 4.0).

**Comment by airspeedswift** [comment]
@swift-ci Please test

**Comment by airspeedswift** [comment]
@swift-ci Please benchmark

**Comment by airspeedswift** [comment]
@swift-ci Please test

**Comment by airspeedswift** [comment]
@swift-ci Please test

File: stdlib/public/SDK/Foundation/NSStringAPI.swift (hunk 1 above)
Comment (moiseev) [inline stdlib/public/SDK/Foundation/NSStringAPI.swift]: Just following the convention in the file.

**Pull request merged**